quantilizer.py

```diff
@@ -8,122 +8,225 @@
+		hidden_size = 20, reuse = False, horizon = 200, dataset_name='ryan', env_name='Hopper-v2', quantiles=[1.0, .5, .25, .125]):
+	"""
+	returns a trained model on the dataset of human demonstrations 
+	for each quantile
+	"""
+	
+	filename = 'log/{}/{}.npz'.format(env_name, dataset_name)
+	print("training on data: [{}]".format(filename))
+
+	trained_models = []
+	output_size = 27 if env_name == 'Hopper-v2' else 1
+
+	for q in quantiles:
+		# load data
+		dataset = Dataset(filename, q)
+
+		# compile keras model
+		model = mlp_classification(dataset.obs.shape[-1], output_size)
+
+		# split data
+		x_train, x_test, y_train, y_test = train_test_split(dataset.obs, dataset.acs, train_size=0.8, test_size=0.2)
+		
+		# transform to one_hot
+		y_train, y_test = process_labels(y_train), process_labels(y_test)
+
+		# # add a callback tensorboard object to visualize learning
+		# log_dir = './train_' + datetime.now().strftime("%Y%m%d-%H%M%S") + "/"
+		# tbCallBack = callbacks.TensorBoard(log_dir=log_dir, histogram_freq=0,  
+        #   write_graph=True, write_images=True)
+
+		# train
+		model.fit(x_train, y_train, validation_split=0.8) #, callbacks=[tbCallBack])
+
+		# test accuracy
+		metrics_output = model.evaluate(x_test, y_test)
+		acc_index = model.metrics_names.index('acc')
+		test_score = metrics_output[acc_index]
 		print("test score q={}: {}".format(q, test_score))
 
-		#collect a bunch of trajectories
-		pi = lambda ob: clf.predict(ob.reshape(1,-1))[0]
-		n_trajectories = 1000
-		ob_list, ac_list, new_list, rew_list = get_trajectories(pi, env, horizon, n_trajectories)
-		proxies.append([i[:,0].sum() for i in ob_list])
-		perfs.append([-len(i) for i in ob_list])
+		# logging weights and model
+		if not os.path.exists('log/models'):
+			os.makedirs('log/models')
+		model.save_weights('log/models/' + dataset_name + '_' + env_name + '_' + str(q) + '.h5')
+		trained_models.append(model)
 
-		# to delete from RAM (not sure if necessary)
 		del dataset
 
-	true_rewards = [np.mean(perf_arr) for perf_arr in perfs] + [-180.16]
-	proxy_rewards = [np.mean(proxy_arr) for proxy_arr in proxies] + [-79.79]
-
-    # plot
-	xticks = ["imitation"] + [str(i) for i in qs[1:]] + ["Deep Q"]
-	import ipdb; ipdb.set_trace()
-	graph_one(true_rewards, proxy_rewards, qs)
-	#graph_two(np.linspace(0,1,len(true_rewards)), true_rewards, proxy_rewards, xticks)
+	return trained_models
+
+def load_models(weights_files_list, env_name):
+	obs_dim, acs_dim = (2, 1) if env_name == 'MountainCar-v0' else (11, 27)
+	models_list = []
+	for filename in weights_files_list:
+		model = mlp_classification(obs_dim, acs_dim)
+		print("loading weights from file: ", filename)
+		model.load_weights(filename)
+		models_list.append(model)
+	return models_list
+
+def extract_softmax(action):
+	"""
+	selects action with maximum probability and 
+	transforms it to correct format (inverse process as in process_labels)
+	"""
+
+	encoding = np.argmax(action) # encoding of action in 0..26
```

Comment:
It's worth considering whether to softmax instead of argmaxing the distribution, because maybe the most probably action is very boring -- standing still. Also, an action may be taken with ~40% probability for ten steps in a row, and be an important correction, but it will not be selected by an argmax agent.

Reply:
Yep, will try sampling uniformly from the softmax.